File: sdk/python/examples/streaming.py
"""
SIGMAX SDK Streaming Example.

Demonstrates real-time streaming analysis with Server-Sent Events (SSE).
Shows progress updates as analysis runs.
"""

import asyncio
from datetime import datetime

from sigmax_sdk import RiskProfile, SigmaxAPIError, SigmaxClient, TradeMode


async def stream_analysis_with_progress(
    client: SigmaxClient,
    symbol: str,
    risk_profile: RiskProfile = RiskProfile.MODERATE,
) -> None:
    """
    Stream analysis for a symbol and display real-time progress.

    Args:
        client: SIGMAX client instance
        symbol: Trading pair to analyze
        risk_profile: Risk profile for analysis
    """
    print(f"\n{'='*60}")
    print(f"Streaming Analysis: {symbol}")
    print(f"Risk Profile: {risk_profile.value}")
    print(f"Started: {datetime.now().strftime('%H:%M:%S')}")
    print(f"{'='*60}\n")

    event_count = 0
    start_time = datetime.now()

    try:
        async for event in client.analyze_stream(
            symbol=symbol,
            risk_profile=risk_profile,
            mode=TradeMode.PAPER,
        ):
            event_count += 1
            event_type = event.get("type", "unknown")
            status = event.get("status", "")
            message = event.get("message", "")

            # Display event
            timestamp = datetime.now().strftime("%H:%M:%S.%f")[:-3]
            print(f"[{timestamp}] Event #{event_count}")
            print(f"  Type: {event_type}")
            if status:
                print(f"  Status: {status}")
            if message:
                print(f"  Message: {message}")

            # Show additional data if present
            if "data" in event:
                data = event["data"]
                if isinstance(data, dict):
                    for key, value in data.items():
                        if key not in ["type", "status", "message"]:
                            print(f"  {key}: {value}")

            # Check if final event
            if event.get("final") or event_type == "complete":
                print("\n✓ Analysis complete!")
                break

            print()  # Blank line between events

    except SigmaxAPIError as e:
        print(f"\n✗ Stream failed: {e.message}")
        if e.status_code:
            print(f"  Status code: {e.status_code}")

    elapsed = (datetime.now() - start_time).total_seconds()
    print(f"\n{'='*60}")
    print(f"Total events: {event_count}")
    print(f"Duration: {elapsed:.2f} seconds")
    print(f"{'='*60}")


async def compare_multiple_streams() -> None:
    """Stream analysis for multiple symbols concurrently."""
    async with SigmaxClient(api_url="http://localhost:8000") as client:
        print("\n" + "=" * 60)
        print("Concurrent Streaming Analysis")
        print("=" * 60)

        # Define symbols to analyze
        symbols = [
            ("BTC/USDT", RiskProfile.CONSERVATIVE),
            ("ETH/USDT", RiskProfile.MODERATE),
            ("SOL/USDT", RiskProfile.AGGRESSIVE),
        ]

        # Create tasks for concurrent streaming
        tasks = [
            stream_analysis_with_progress(client, symbol, risk)
            for symbol, risk in symbols
        ]

        # Run all streams concurrently
        await asyncio.gather(*tasks, return_exceptions=True)


async def single_stream_example() -> None:
    """Simple single-stream analysis example."""
    async with SigmaxClient(api_url="http://localhost:8000") as client:
        await stream_analysis_with_progress(
            client,
            "BTC/USDT",
            RiskProfile.MODERATE,
        )


async def filtered_stream_example() -> None:
    """Filter and process specific event types."""
    async with SigmaxClient(api_url="http://localhost:8000") as client:
        print("\n" + "=" * 60)
        print("Filtered Streaming (status events only)")
        print("=" * 60 + "\n")

        async for event in client.analyze_stream(
            "BTC/USDT",
            RiskProfile.MODERATE,
        ):
            # Only show status update events
            if event.get("type") == "status":
                timestamp = datetime.now().strftime("%H:%M:%S")
                status = event.get("status", "unknown")
                message = event.get("message", "")
                print(f"[{timestamp}] {status}: {message}")

            # Stop on completion
            if event.get("final") or event.get("type") == "complete":
                print("\nAnalysis finished!")
                break


async def main() -> None:
    """Run streaming examples."""
    print("\n" + "=" * 60)
    print("SIGMAX SDK - Streaming Examples")
    print("=" * 60)

    try:
        # Example 1: Single stream with full details
        print("\n\n=== Example 1: Single Stream ===")
        await single_stream_example()

        # Example 2: Filtered stream (status updates only)
        print("\n\n=== Example 2: Filtered Stream ===")
        await filtered_stream_example()

        # Uncomment to run concurrent streams (may be noisy)
        # print("\n\n=== Example 3: Concurrent Streams ===")
        # await compare_multiple_streams()

    except KeyboardInterrupt:
        print("\n\nStream interrupted by user")
    except Exception as e:
        print(f"\n\nUnexpected error: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/typescript/examples/streaming.ts
/**
 * SIGMAX SDK - Streaming Analysis Example
 *
 * This example demonstrates real-time streaming analysis with:
 * - Server-Sent Events (SSE)
 * - Step-by-step agent updates
 * - Real-time decision making
 * - Async iteration pattern
 */

import { SigmaxClient, StreamEvent } from '@sigmax/sdk';

async function main() {
  // Initialize the client
  const client = new SigmaxClient({
    apiKey: process.env.SIGMAX_API_KEY || 'your-api-key-here',
    apiUrl: process.env.SIGMAX_API_URL || 'http://localhost:8000',
  });

  const symbol = 'ETH/USDT';

  console.log('🚀 SIGMAX SDK - Streaming Analysis Example\n');
  console.log(`📡 Starting live analysis stream for ${symbol}...\n`);
  console.log('━'.repeat(60));

  try {
    let sessionId: string | undefined;
    let stepCount = 0;

    // Stream analysis with real-time updates
    for await (const event of client.analyzeStream(symbol, {
      risk_profile: 'balanced',
      mode: 'paper',
    })) {
      handleStreamEvent(event);

      // Track session
      if (event.type === 'meta' && event.session_id) {
        sessionId = event.session_id;
      }

      // Count steps
      if (event.type === 'step') {
        stepCount++;
      }

      // Exit on final or error
      if (event.type === 'final' || event.type === 'error') {
        break;
      }
    }

    console.log('\n' + '━'.repeat(60));
    console.log(`\n✅ Analysis completed!`);
    console.log(`Session ID: ${sessionId}`);
    console.log(`Total steps: ${stepCount}\n`);

  } catch (error) {
    console.error('\n❌ Stream error:', error);
    process.exit(1);
  }
}

/**
 * Handle different stream event types
 */
function handleStreamEvent(event: StreamEvent): void {
  const timestamp = new Date(event.timestamp).toLocaleTimeString();

  switch (event.type) {
    case 'meta':
      console.log(`\n[${timestamp}] 🔗 Connected to stream`);
      console.log(`Session: ${event.session_id}`);
      break;

    case 'step':
      console.log(`\n[${timestamp}] 🔄 ${formatStepName(event.step || 'Unknown')}`);
      if (event.update) {
        console.log(`  ${formatUpdate(event.update)}`);
      }
      break;

    case 'final':
      console.log(`\n[${timestamp}] ✨ FINAL DECISION`);
      console.log('━'.repeat(60));

      if (event.decision) {
        console.log(`\nAction: ${formatAction(event.decision.action)}`);
        console.log(`Confidence: ${formatConfidence(event.decision.confidence)}`);
        console.log(`Reasoning: ${event.decision.reasoning || 'N/A'}`);
      }

      if (event.proposal) {
        console.log(`\n💡 TRADE PROPOSAL`);
        console.log(`Proposal ID: ${event.proposal.proposal_id}`);
        console.log(`Symbol: ${event.proposal.symbol}`);
        console.log(`Action: ${formatAction(event.proposal.action)}`);
        console.log(`Size: ${event.proposal.size}`);
        console.log(`Notional: $${event.proposal.notional_usd.toFixed(2)}`);
        console.log(`Rationale: ${event.proposal.rationale || 'N/A'}`);
      }

      if (event.state) {
        console.log(`\n📊 STATE`);
        console.log(JSON.stringify(event.state, null, 2));
      }
      break;

    case 'error':
      console.error(`\n[${timestamp}] ❌ ERROR: ${event.error}`);
      break;

    default:
      console.log(`\n[${timestamp}] 📦 ${event.type}:`, event.data);
  }
}

/**
 * Format step name for display
 */
function formatStepName(step: string): string {
  const stepMap: Record<string, string> = {
    research: '🔍 RESEARCH',
    bull_debate: '🐂 BULL CASE',
    bear_debate: '🐻 BEAR CASE',
    analyzer: '📊 TECHNICAL ANALYSIS',
    risk: '🛡️ RISK VALIDATION',
    decision: '⚖️ FINAL DECISION',
  };

  return stepMap[step] || step.toUpperCase();
}

/**
 * Format update message
 */
function formatUpdate(update: any): string {
  if (typeof update === 'string') {
    return update;
  }

  if (typeof update === 'object') {
    // Extract key information
    const parts: string[] = [];

    if (update.message) parts.push(update.message);
    if (update.status) parts.push(`Status: ${update.status}`);
    if (update.progress !== undefined) parts.push(`Progress: ${update.progress}%`);

    return parts.length > 0 ? parts.join(' | ') : JSON.stringify(update);
  }

  return String(update);
}

/**
 * Format action with emoji
 */
function formatAction(action: string): string {
  const actionMap: Record<string, string> = {
    buy: '🟢 BUY',
    sell: '🔴 SELL',
    hold: '🟡 HOLD',
  };

  return actionMap[action.toLowerCase()] || action.toUpperCase();
}

/**
 * Format confidence as percentage with bar
 */
function formatConfidence(confidence: number): string {
  const percent = (confidence * 100).toFixed(1);
  const barLength = 20;
  const filled = Math.round(confidence * barLength);
  const bar = '█'.repeat(filled) + '░'.repeat(barLength - filled);

  return `${percent}% ${bar}`;
}

// Run the example
main();
